### PR TITLE
Feature/fixing clang gnu compiler warnings

### PR DESCRIPTION
**Pull request by phelter**
Fixing Clang and GNU compiler warnings along with some clang-tidy issues.

Description
-----------
* Upkeep/maintenance - fixed a bunch of the compiler warnings that were previously ignored. See lists in CMakeLists.txt for ones fixed.

Test Steps
-----------
* None - re-run test suite.
* 
Checklist:
----------
- [x ] I have tested my changes. No regression in existing tests.
- [x ] I have modified and/or added unit-tests to cover the code changes in this Pull Request.

Related Issue
-----------
None.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: include/ff_error.h [hunk 1]
```diff
@@ -62,6 +62,7 @@
 #define FF_GETMOD_FUNC( x )    ( ( ( ( unsigned ) x ) >> FF_FUNCTION_SHIFT ) & 0xFFFF )
 #define FF_ERRFLAG    0x80000000
 #define FF_isERR( x )          ( ( ( x ) & FF_ERRFLAG ) != 0 )
+#define FF_createERR( x, id ) ( ( FF_Error_t ) ( ( x ) |  ( id ) ) )
```
Comment (phelter) [inline include/ff_error.h:65]: Note - added `FF_createERR` to ensure the errors are always created the same way.  This allows easier updates if at some point in the future want to ensure x  and id are a particular enumeration type.

File: CMakeLists.txt [hunk 2]
```diff
@@ -186,22 +186,16 @@ target_compile_definitions( freertos_plus_fat
 
 target_compile_options( freertos_plus_fat
   PRIVATE
+    $<$<COMPILE_LANG_AND_ID:C,GNU>:-Wno-array-bounds>
```
Comment (phelter) [inline CMakeLists.txt:189]: This is compiling with clang 14.0

Comment (dachalco) [reply CMakeLists.txt:189]: Nit: I feel array-bounds violations are fairly useful warnings, but I'm unaware whether there were intended "false positive" types being reported. Regardless plenty of other warnings were fixed, so approved first round of reviews.

Comment (dachalco) [reply CMakeLists.txt:189]: I built POSIX test on Ubuntu 22.04 with GCC11 with `Warray-bounds` and no array bound warnings came up.

**Review by dachalco — DISMISSED** [review]

**Comment by dachalco** [comment]
@phelter 
I tried to push a commit to fix the failing formatting checks but did not have permissions for this PR.
To fix the failing formatting checks, the [failing files](https://github.com/FreeRTOS/Lab-Project-FreeRTOS-FAT/actions/runs/4854835580/jobs/8652691066?pr=45) need to be [uncrustified](https://github.com/uncrustify/uncrustify) (v0.69).

File: ff_format.c [hunk 3]
```diff
@@ -252,13 +252,13 @@ static FF_Error_t prvFormatGetClusterSize( struct xFormatSet * pxSet,
 
     for( ; ; )
     {
-        int32_t groupSize;
+        uint32_t groupSize;
```
Comment (htibosch) [inline ff_format.c:255]: Change it to `ulGroupSize`?

File: ff_ioman.c [hunk 4]
```diff
@@ -698,7 +690,7 @@ int32_t FF_BlockWrite( FF_IOManager_t * pxIOManager,
  * It may be set to one of these values: FF_T_FAT[12,16,32]
  * just to force the driver to assume a certain FAT type.
  */
-uint8_t ucAssumeFATType;
+static uint8_t ucAssumeFATType = 0;
```
Comment (htibosch) [inline ff_ioman.c:693]: Unless we provide an access function to change `ucAssumeFATType`, I'd like to keep it global.

File: ff_stdio.c [hunk 5]
```diff
@@ -2179,9 +2178,7 @@ int prvFFErrorToErrno( FF_Error_t xError )
         xTime.tm_mon = pxTime->Month - 1;
         xTime.tm_year = pxTime->Year - 1900;
 
-        xReturn = FreeRTOS_mktime( &xTime );
-
-        return xReturn;
+        return FreeRTOS_mktime( &xTime );
```
Comment (htibosch) [inline ff_stdio.c:2181]: Here I prefer the two steps: assign to a variable and return it, so it can be inspected while debugging.
An optimising compiler will remove the extra step.

File: ff_stdio.c [hunk 6]
```diff
@@ -2167,10 +2167,9 @@ int prvFFErrorToErrno( FF_Error_t xError )
 
 #if ( ffconfigTIME_SUPPORT == 1 )
 
-    static uint32_t prvFileTime( FF_SystemTime_t * pxTime )
+    time_t prvFileTime( FF_SystemTime_t * pxTime )
```
Comment (htibosch) [inline ff_stdio.c:2170]: Here above `prvFileTime()` was declared private. I don't think that the function is useful outside this module.

File: ff_time.c [hunk 7]
```diff
@@ -105,8 +104,8 @@
     #define WEEK_DAY_SATURDAY       6
 
 /* Make a bitmask with a '1' for each 31-day month. */
-    #define _MM( month )    ( 1u << ( month - 1 ) )
-    #define MASK_LONG_MONTHS     ( _MM( 1 ) | _MM( 3 ) | _MM( 5 ) | _MM( 7 ) | _MM( 8 ) | _MM( 10 ) | _MM( 12 ) )
+    #define MM_( month )    ( 1u << ( month - 1 ) )
```
Comment (htibosch) [inline ff_time.c:107]: Just curious: why this change: from `_MM` to `MM_`? Did you run MISRA checks?

File: ff_time.c [hunk 8]
```diff
@@ -141,7 +140,7 @@
 
     static portINLINE unsigned long ulDaysPerYear( int iYear )
```
Comment (htibosch) [inline ff_time.c:141]: Shall we replace all occurrences of `unsigned long` in this module with `uint32_t`?
MISRA would complain about the ambiguity of `unsigned long`.

File: include/ff_stdio.h [hunk 9]
```diff
@@ -87,10 +87,11 @@
         uint16_t st_dev;  /* The device on which the file can be found (see ff_sys.c) */
         uint16_t st_mode; /* The mode (attribute bits) of this file or directory. */
 
+        /* Note time_t must be used here otherwise will have bugs when 2032 (uint32_t second clock rolls over) */
```
Comment (htibosch) [inline include/ff_stdio.h:90]: Yes good, thank you!

File: include/ff_error.h (hunk 1 above)
Comment (htibosch) [reply include/ff_error.h:65]: That's a good choice, thank you

File: portable/ATSAM4E/ff_sddisk.c [hunk 10]
```diff
@@ -284,14 +284,14 @@ static int32_t prvFFWrite( uint8_t * pucBuffer,
...
-            FF_PRINTF( "prvFFWrite: %lu: %ld\n", ulSectorNumber, lResult );
+            FF_PRINTF( "prvFFWrite: %u: %d\n", (unsigned) ulSectorNumber, (int) lResult );
         }
     }
     else
     {
         if( pxDisk->xStatus.bIsInitialised != pdFALSE )
         {
-            FF_PRINTF( "prvFFWrite: warning: %lu + %lu > %lu\n", ulSectorNumber, ulSectorCount, pxDisk->ulNumberOfSectors );
+            FF_PRINTF( "prvFFWrite: warning: %u + %u > %u\n", (unsigned) ulSectorNumber, (unsigned)ulSectorCount, (unsigned)pxDisk->ulNumberOfSectors );
```
Comment (htibosch) [inline portable/ATSAM4E/ff_sddisk.c:294]: Casting to `int` and `unsigned` is also my preference while logging, thank you.

**Review by htibosch — APPROVED** [review]
Hello @phelter, you did a very good job!

One more comment about a line that is not visible here below:
~~~diff
-    if( uxIndex == 0 )
+    if( uxIndex == 0U )
     {
-         uxIndex = 1;
+         uxIndex = 1U;
     }
~~~
that is because you made `xIndex` unsigned:
~~~diff
-    BaseType_t xIndex;
+    UBaseType_t uxIndex;
~~~
I approve the changes, thank you.

**Comment by htibosch** [comment]
Why not create a new PR which will "uncrustify" the source code?

File: ff_stdio.c (hunk 5 above)
Comment (phelter) [reply ff_stdio.c:2181]: Not all compilers will remove the extra step especially if the same value is used above.  And if you want to look at the value you can just break on the `return` line, enter the `FreeRTOS_mktime()`.  and check the value there.  Or where it is consumed in the function above.

File: ff_time.c (hunk 7 above)
Comment (phelter) [reply ff_time.c:107]: @htibosch - macros and precompiler definitions that start with `_` are considered reserved and should not be used.
https://clang.llvm.org/docs/DiagnosticsReference.html#wreserved-identifier

For MISRA checks - if they were run in the ci.yml then they pass.  If not then I suggest if this is a requirement then you should add it to the ci.yml to check that.

File: ff_time.c (hunk 8 above)
Comment (phelter) [reply ff_time.c:141]: Yes I would suggest that, but out of scope of this particular PR to reduce those specific compiler warnings.

Also using `size_t` and `uintptr_t` for sizes and addresses is also best.

File: ff_format.c (hunk 3 above)
Comment (phelter) [reply ff_format.c:255]: feel free to change once this has been updated.  I truly don't see the point in adding superfluous prefixes to values to identify the type when the IDE's these days can tell you that with pop-ups or other features.  My 2 cents but if that is the format or flow then I suggest you add a tool to check and fix those for you since it is a simple thing to do automatically.
https://clang.llvm.org/docs/ClangFormatStyleOptions.html

I understand you use uncrustify for this process, but it seems as though the bots for doing that aren't working.

File: ff_ioman.c (hunk 4 above)
Comment (phelter) [reply ff_ioman.c:693]: @htibosch - ucAssumeFATType is currently not global.  If you would like it global then you need to define it in a header somewhere.  Eg:

```c
extern uint8_t ucAssumeFATType;
```

But that is not the case right now.  Considering an issue that you can resolve outside this PR.

File: portable/STM32F4xx/ff_sddisk.c [hunk 11]
```diff
@@ -866,11 +866,11 @@ static BaseType_t prvSDMMCInit( BaseType_t xDriveNumber )
             }
         }
     #endif /* if ( BUS_4BITS != 0 ) */
-    FF_PRINTF( "HAL_SD_Init: %d: %s type: %s Capacity: %lu MB\n",
+    FF_PRINTF( "HAL_SD_Init: %d: %s type: %s Capacity: %u MB\n",
                SD_state,
                prvSDCodePrintable( ( uint32_t ) SD_state ),
                xSDHandle.CardType == HIGH_CAPACITY_SD_CARD ? "SDHC" : "SD",
-               xSDCardInfo.CardCapacity / ( 1024 * 1024 ) );
+               (unsigned)(xSDCardInfo.CardCapacity / ( 1024 * 1024 ) ));
```
[the remark this answers is not in the document]
Comment (phelter) [reply portable/STM32F4xx/ff_sddisk.c:873]: Yes that could be done in lieu of this change.

File: ff_file.c [hunk 12]
```diff
@@ -566,16 +562,15 @@ static FF_FILE * prvAllocFileHandle( FF_IOManager_t * pxIOManager,
...
-        BaseType_t pathLen = STRLEN( pcPath );
+        UBaseType_t pathLen = STRLEN( pcPath );
 
         FF_PendSemaphore( pxIOManager->pvSemaphore );
         {
-            for( xIndex = 0; xIndex < ffconfigPATH_CACHE_DEPTH; xIndex++ )
+            for( UBaseType_t xIndex = 0; xIndex < ffconfigPATH_CACHE_DEPTH; xIndex++ )
             {
-                BaseType_t len2 = STRLEN( pxIOManager->xPartition.pxPathCache[ xIndex ].pcPath );
+                UBaseType_t len2 = STRLEN( pxIOManager->xPartition.pxPathCache[ xIndex ].pcPath );
```
[the remark this answers is not in the document]
Comment (phelter) [reply ff_file.c:571]: Goal was not to fix the names of variables to match their types.  Goal was to use the correct types.

**Review by phelter — COMMENTED** [review]
The ci build bot for performing uncrustify is broken when merging from a fork.  Please provide the proper method for executing uncrustify, or fix the bot.

I have no intention on installing uncrustify as I consider it a subset of the tools that I already use (clang-format) and have no use for it other than to fix code in freertos.

**Comment by phelter** [comment]
@moninom1 
> /bot run uncrustify

The bot is not working I've tried, and so have others.  I don't know if it has something to do with this being a merge from a fork?

**Comment by moninom1** [comment]
@phelter Yeah i checked. Adding changes to fix the issue in a separate PR. Will update once merged.

**Review by moninom1 — APPROVED** [review]

File: test/build-combination/Common/main.c [hunk 13]
```diff
@@ -35,13 +35,11 @@
 #include "task.h"
 
 /* System application includes. */
+#include "FreeRTOSConfig.h"
```
Comment (aggarg) [inline test/build-combination/Common/main.c:38]: This should not be needed explicitly - including "FreeRTOS.h" is enough.

**Review by aggarg — APPROVED** [review]

File: test/build-combination/Common/main.c (hunk 13 above)
Comment (htibosch) [reply test/build-combination/Common/main.c:38]: +1
The first thing FreeRTOS.h does is :
~~~
/* Application specific configuration options. */
#include "FreeRTOSConfig.h"
~~~

**Review by kar-rahul-aws — APPROVED** [review]
